source/common/ur_util.hpp

```diff
@@ -266,6 +266,59 @@ inline ur_result_t exceptionToResult(std::exception_ptr eptr) {
 
 template <class> inline constexpr bool ur_always_false_t = false;
 
+namespace {
+// Compile-time map, mapping a UR list node type, to the enum tag type
+// These are helpers for the `find_stype_node` helper below
+template <ur_structure_type_t val> struct stype_map_impl {
+    static constexpr ur_structure_type_t value = val;
+};
+
+template <typename T> struct stype_map {};
+// contains definitions of the map specializations e.g.
+// template <> struct stype_map<ur_usm_device_desc_t> :
+// stype_map_impl<UR_STRUCTURE_TYPE_USM_DEVICE_DESC> {};
+#include "stype_map_helpers.def"
+
+template <typename T> constexpr int as_stype() { return stype_map<T>::value; };
+
+/// Walk a generic UR linked list looking for a node of the given type. If it's
+/// found, its address is returned, othewise `nullptr`. e.g. to find out whether
+/// a `ur_usm_host_desc_t` exists in the given polymorphic list, `mylist`:
+///
+/// ```cpp
+/// auto *node = find_stype_node<ur_usm_host_desc_t>(mylist);
+/// if (!node)
+///   printf("node of expected type not found!\n");
+/// ```
+///
+/// To find multiple nodes of a given type:
+///
+/// ```cpp
+/// std::vector<ur_usm_host_desc_t> nodes;
+///
+/// for (auto *node = find_stype_node<ur_usm_host_desc_t>(mylist); node; node =
```

Comment:
Doesn't this assume that all the elements following the first `ur_usm_host_desc_t` are of that same type?

Reply:
This snippet is an example which finds all `ur_usm_host_desc_t` in a linked list. It's artificial and probably would never occur in real code. The linked list in question can have nodes of any type

Reply:
Right, but the comment and the code is wrong, because, as you note, the list can have nodes of any types. So, if I read this right, this will find the first node of the type `ur_usm_host_desc_t` and then assume everything that comes after is is also `ur_usm_host_desc_t`. Or I'm missing something?

Reply:
e.g.
in the following imaginary list
```
(a) ur_usm_alloc_location_desc_t -> (b )ur_usm_host_desc_t -> (c) ur_usm_device_desc_t -> (d) ur_usm_host_desc_t
```

The example loop above will push (b) and (d) onto the `nodes` vector

Reply:
I see what you're saying.
I guess there technically needs to be a cast in the example, but given how these lists work in practise, I don't think we need to care about type-safety in example code

Reply:
> e.g. in the following imaginary list
> 
> ```
> (a) ur_usm_alloc_location_desc_t -> (b )ur_usm_host_desc_t -> (c) ur_usm_device_desc_t -> (d) ur_usm_host_desc_t
> ```
> 
> The example loop above will push (b) and (d) onto the `nodes` vector

I don't think that's true. It would push (b), (c) an (d) onto the vector, incorrectly casting (c) to `ur_usm_host_desc_t`.

Reply:
I see the bug!
+/// ```cpp
+/// std::vector<ur_usm_host_desc_t> nodes;
+///
+/// for (ur_usm_host_desc_t *node; node = find_stype_node<ur_usm_host_desc_t>(mylist); node; node = static_cast<ur_usm_host_desc_t>(node->pNext)) {
+///   nodes.push_back(node);
+/// }

Reply:
this would be an infinite loop because you never update mylist ;)

Maybe this would be simpler:
```
auto *node = mylist;
while (node = find_stype_node<ur_usm_host_desc_t>(node)) {
  ...
  node = node->pNext; // potentially casts incorrectly from void *. Not sure if this isn't UB.
}
```

This is surprisingly tricky given the `find_stype_node` as written :D

Reply:
OH jeez. Perhaps I should just remove this example, seeing as I can't get it right, and it's a very unlikely use-case?

Reply:
Sure, that's the simplest solution ;)

Reply:
My brain isn't working this morning, clearly. Thanks for sticking with me!